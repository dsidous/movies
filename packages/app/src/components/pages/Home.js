import { compose, withProps } from 'recompose';
import { withUpcomingMovies } from '@movies/common';

import Home from '../templates/Home';

export default compose(
  withProps(() => ({ query: 'page=1' })),
  withUpcomingMovies(),
)(Home);
