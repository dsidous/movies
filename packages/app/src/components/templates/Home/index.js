import React from 'react';
import {
  Text,
  FlatList,
  View,
  Dimensions,
  Image,
  StyleSheet,
  Animated,
  Platform,
} from 'react-native';
import { StatusBar } from 'expo-status-bar';
import { LinearGradient } from 'expo-linear-gradient';

const getImagePath = path =>
  `https://image.tmdb.org/t/p/w440_and_h660_face${path}`;

const getBackdropPath = path =>
  `https://image.tmdb.org/t/p/w370_and_h556_multi_faces${path}`;

const { width, height } = Dimensions.get('window');

const SPACING = 10;
const ITEM_SIZE = Platform.OS === 'ios' ? width * 0.72 : width * 0.74;
const SPACER_ITEM_SIZE = (width - ITEM_SIZE) / 2;
const BACKDROP_HEIGHT = height * 0.65;

const styles = StyleSheet.create({
  posterImage: {
    width: '100%',
    height: ITEM_SIZE * 1.2,
    resizeMode: 'cover',
    borderRadius: 24,
  },
});

const Backdrop = ({ movies, scrollX }) => {
  return (
    <View
      style={{
        position: 'absolute',
        width,
        height: BACKDROP_HEIGHT,
      }}
    >
      <FlatList
        data={movies}
        keyExtractor={item => `${item.id}-backdrop`}
        removeClippedSubviews={false}
        contentContainerStyle={{ width, height: BACKDROP_HEIGHT }}
        renderItem={({ item, index }) => {
          if (!item.backdrop_path) {
            return null;
          }
          const translateX = scrollX.interpolate({
            inputRange: [(index - 2) * ITEM_SIZE, (index - 1) * ITEM_SIZE],
            outputRange: [0, width],
            // extrapolate:'clamp'
          });
          return (
            <Animated.View
              removeClippedSubviews={false}
              style={{
                position: 'absolute',
                width: translateX,
                height,
                overflow: 'hidden',
              }}
            >
              <Image
                source={{ uri: getBackdropPath(item.backdrop_path) }}
                style={{
                  width,
                  height: BACKDROP_HEIGHT,
                  position: 'absolute',
                }}
              />
            </Animated.View>
          );
        }}
      />
      <LinearGradient
        colors={['rgba(0, 0, 0, 0)', 'white']}
        style={{
          height: BACKDROP_HEIGHT,
          width,
          position: 'absolute',
          bottom: 0,
        }}
      />
    </View>
  );
};

const Home = ({ upcomingLoading, upcoming }) => {
  const scrollX = React.useRef(new Animated.Value(0)).current;

  if (upcomingLoading) {
    return <Text>Loading...</Text>;
  }

  const movies = [{ id: 'left-spacer' }, ...upcoming, { id: 'right-spacer' }];

  return (
    <View style={{ flex: 1 }}>
      <Backdrop movies={movies} scrollX={scrollX} />
      <StatusBar hidden />
      <Animated.FlatList
        showsHorizontalScrollIndicator={false}
        data={movies}
        keyExtractor={item => item.id}
        horizontal
        contentContainerStyle={{
          alignItems: 'center',
        }}
        snapToInterval={ITEM_SIZE}
        snapToAlignment="start"
        decelerationRate={Platform.OS === 'ios' ? 0 : 0.98}
        bounces={false}
        onScroll={Animated.event(
          [{ nativeEvent: { contentOffset: { x: scrollX } } }],
          { useNativeDriver: false },
        )}
        scrollEventThrottle={16}
        renderItem={({ item, index }) => {
          if (!item.poster_path) {
            return (
              <View
                style={{
                  width: SPACER_ITEM_SIZE,
                }}
              />
            );
          }

          const inputRange = [
            (index - 2) * ITEM_SIZE,
            (index - 1) * ITEM_SIZE,
            index * ITEM_SIZE,
          ];

          const translateY = scrollX.interpolate({
            inputRange,
            outputRange: [100, 50, 100],
          });

          return (
            <View
              style={{
                width: ITEM_SIZE,
              }}
            >
              <Animated.View
                style={{
                  marginHorizontal: SPACING,
                  padding: SPACING * 2,
                  alignItems: 'center',
                  backgroundColor: 'white',
                  borderRadius: 24,
                  transform: [{ translateY }],
                }}
              >
                <Image
                  source={{ uri: getImagePath(item.poster_path) }}
                  style={styles.posterImage}
                />
                <Text>{item.title}</Text>
              </Animated.View>
            </View>
          );
        }}
      />
    </View>
  );
};

export default Home;
