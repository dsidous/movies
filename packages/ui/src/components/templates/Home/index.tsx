/** @jsxFrag React.Fragment */
import React from 'react';

import PageTransition from '../../atoms/PageTransition';
import Section from '../../atoms/Section';
import TopRatedMovies from '../../organisms/TopRatedMovies';
import SlideList from '../../molecules/SlideList';
import { Movie } from '@typesRoots/movie';

interface Props {
  nowPlayingLoading: boolean;
  upcomingLoading: boolean;
  nowplaying: Movie[];
  upcoming: Movie[];
  popular: Movie[];
}

const Home: React.FC<Props> = ({
  nowPlayingLoading,
  upcomingLoading,
  nowplaying,
  upcoming,
  popular,
}) => {
  return (
    <PageTransition>
      <TopRatedMovies popular={popular} />

      {!nowPlayingLoading && !upcomingLoading && (
        <>
          <Section noborder disableGutters>
            <Section.Header>In Theatres</Section.Header>
            <SlideList items={nowplaying} />
          </Section>
          <Section noborder disableGutters>
            <Section.Header>Upcoming Movies</Section.Header>
            <SlideList items={upcoming} />
          </Section>
        </>
      )}
    </PageTransition>
  );
};

export default Home;
