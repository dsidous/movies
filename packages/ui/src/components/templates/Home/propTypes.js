import { array, bool, object, oneOfType } from 'prop-types';

export const propTypes = {
  popular: array,
  nowPlayingLoading: bool,
  popularLoading: bool,
  upcomingLoading: bool,
  nowplaying: oneOfType([object, array]),
  upcoming: oneOfType([object, array]),
};

export const defaultProps = {
  nowplaying: [],
  upcoming: [],
};
