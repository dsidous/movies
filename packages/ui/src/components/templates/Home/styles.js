import { makeStyles } from '@material-ui/core/styles';

export default makeStyles(
  theme => ({
    root: {
      display: 'grid',

      [theme.breakpoints.up('sm')]: {
        gridTemplateColumns: '1fr 1fr',
        margin: '0 auto',
        width: '70%    ',
      },
    },
  }),
  { name: 'topList' },
);
