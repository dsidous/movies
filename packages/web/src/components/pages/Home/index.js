import { compose, withProps } from 'recompose';
import { Home } from '@movies/ui';
import {
  withPopularMovies,
  withNowPlayingMovies,
  withUpcomingMovies,
} from '@movies/common';

export default compose(
  withProps(() => ({ query: 'page=1' })),
  withPopularMovies(),
  withNowPlayingMovies(),
  withUpcomingMovies(),
)(Home);
