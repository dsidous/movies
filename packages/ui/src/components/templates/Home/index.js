import React from 'react';

import { propTypes, defaultProps } from './propTypes';
import Spinner from '../../atoms/Spinner';
import PageTransition from '../../atoms/PageTransition';
import TopRatedMovies from '../../organisms/TopRatedMovies';
import HomeList from '../../atoms/HomeList';

const Home = ({
  nowPlayingLoading,
  popularLoading,
  upcomingLoading,
  nowplaying,
  upcoming,
  popular,
}) => {
  if (nowPlayingLoading || popularLoading || upcomingLoading) {
    return <Spinner />;
  }

  return (
    <PageTransition>
      <TopRatedMovies popular={popular} />

      <div className="top-lists-wrapper">
        <HomeList list={nowplaying.slice(0, 10)} title="In Theatres" />
        <HomeList list={upcoming.slice(0, 10)} title="Upcoming Movies" />
      </div>
    </PageTransition>
  );
};

export default Home;

Home.defaultProps = defaultProps;

Home.propTypes = propTypes;
