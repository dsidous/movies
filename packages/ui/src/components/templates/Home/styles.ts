import { makeStyles } from '@material-ui/core/styles';

export default makeStyles(
  theme => ({
    root: {
      [theme.breakpoints.up('sm')]: {
        margin: '0 auto',
      },
    },
  }),
  { name: 'topList' },
);
