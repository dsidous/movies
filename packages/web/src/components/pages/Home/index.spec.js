import React from 'react';
import { shallow } from 'enzyme';
import Home from '.';

describe('Pages/Home', () => {
  it('should render as expected', () => {
    const wrapper = shallow(<Home />);

    expect(wrapper).toMatchSnapshot();
  });
});
