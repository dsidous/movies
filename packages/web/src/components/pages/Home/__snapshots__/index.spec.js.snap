// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages/Home should render as expected 1`] = `ShallowWrapper {}`;
